### ISSUE TITLE
Bazel support

### ISSUE THREAD

**Issue by softprops** [scalameta/metals-feature-requests#39]
**Is your feature request related to a problem? Please describe.**

As a vscode user for scala project's build with bazel I'd like to see scala meta support for bazel become more of a first class citizen.

**Describe the solution you'd like**

I would like to see bazel become a first class build tool for this project.

**Describe alternatives you've considered**

Intellij + it's bazel plugin for scala is another option, ideally I'd like to stick with vscode.

**Additional context**

**Search terms**:

**Comment by tgodzik** [comment]
Hi, thanks for reporting! The Bazel integration is dependent on https://github.com/scalacenter/bloop/issues/427 There is already some preliminary work done and I think bloop configuration can be generated. In case of existing bloop configuration files Metals will pick that up and should work.

For sure we will add automatic integration with Bazel as soon as the integration is more stable.

**Comment by tmccombs** [comment]
Ideally, I'd like to use Metas with bazel without having to use bloop. How feasible is that?

**Comment by kpbochenek** [comment]
Metals require another component for providing build information, running and testing your code.
Communication with this component is via protocol called BSP(build server protocol) and you can see tools implementing it here(bloop/mill):
https://build-server-protocol.github.io/docs/implementations.html
Bloop implements this protocol but you should be able to swap it for any other tool that supports BSP.
If Bazel implements BSP then it could be used instead of bloop.

**Comment by olafurpg** [comment]
@tmccombs Related links if somebody is interested in implementing BSP on top of Bazel

* https://docs.google.com/document/d/1KCQMvUMNVUBDyDmZJ1DB80tf21MVQdyWYS0Y9D_P0as/edit
* https://github.com/illicitonion/bazel-bsp
* https://contributors.scala-lang.org/t/build-server-protocol-in-sbt/4234/17

I think it should be fairly straight-forward to translate Bazel build graphs into BSP. I estimate the biggest challenge is delivering low-latency diagnostic (in particular clearing diagnostics, see https://youtu.be/MRQMylDxBJ8?t=709).

There are currently active efforts to implement BSP in sbt and for Mill, it would be great to have something similar for Bazel!

**Comment by tmccombs** [comment]
Thank you, I'll take a look at that.

**Comment by tmccombs** [comment]
I found another project at https://github.com/JetBrains/bazel-bsp.

However, I'm unclear on how I configure metals to use a custom build server. I specifically use Neovim, but I can't find instructions on how to set up metals to use a different build server for any editor.

**Comment by tgodzik** [comment]
Usually it should be enough to have .bsp/*,json generated with : https://build-server-protocol.github.io/docs/server-discovery.html

**Comment by ckipp01** [comment]
@tmccombs if you're using `coc-metals`, you can switch to the master branch and you'll find there is a new `bsp-switch` command that was recently added which will allow you to do pretty much what you want.

You'll still want to do as @tgodzik mentioned and make sure there is a `.bsp/*.json` created by `bazel-bsp`, but you should be able to connect to it. Also make sure you're on the latest snapshot of Metals if you try this because we just merged in more a pr that will greatly help with the facilitation of switching BSP servers.

**Comment by ex0ns** [comment]
Hello, I would be very interested by this feature, but I have to say that I'm pretty confused right now, could we please try to clarify the situation here:

1. If we want to work with bazel on top of bloop, we are waiting on scalacenter/bloop#427 which did not make any progress (https://github.com/bazelbuild/rules_scala/pull/997) for the past few months
2. We could use bazel-bsp instead of bloop to generate the file .bsp/.json files and then connect to it using Metals ?
 
Am I correct ? Did I forget something ? 

Thanks

**Comment by ckipp01** [comment]
Thanks for prompting the discussion @ex0ns!

As I posted a few days ago, we've recently added in some more genereal BSP support that I think should make the process of using other build servers easier. In in ideal world, but of your situations in 1 and 2 will work. It'd be great to be able to use Bloop to compile a Bazel build, and it's also great to see the work that has gone into [bazel-bsp](https://github.com/JetBrains/bazel-bsp). I think for some more concrete steps I'd say I can outline two:

- I haven't gotten around to having the time to dive into it yet (as I'm not super familiar with Bazel), but I'd really like to give the `bazel-bsp` a go with Metals to see what happens. Seeing that it's already available, I think it's a smart to focus on this from the Metals side to ensure the BSP part is solid.
- Currently we also have support for various build tools to pick up changes in the build definition based on changes in the related files. This doesn't yet exist for Bazel in Metals, so this will also have to be added.

I'd say those are the two concrete steps that would need to go into tackling this feature-request unless the [Bloop integration pr](https://github.com/bazelbuild/rules_scala/pull/997) gets more movement.

As a total aside, I want to reiterate that it's great to see the various build tools supporting BSP, and it also great to see the fantastic work the JetBrains folks put into `bazel-bsp`. This type of work benefits everyone in the community, and it makes me exciting for the future of Scala tooling.

**Comment by ValdemarGr** [comment]
Before this is ready, I use this to generate a configuration.
https://gist.github.com/ValdemarGr/fbf2f574d12da0c23513b581d095154f
Keep in mind that I use https://github.com/bazelbuild/rules_jvm_external

**Comment by ex0ns** [comment]
I gave bazel-bsp a try, for that I had to 
I created a new project with a single file to test, here is the content (I can create a repo if you want, but it's really small)

```
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "bazel_skylib",
    urls = [
        "https://github.com/bazelbuild/bazel-skylib/releases/download/1.0.3/bazel-skylib-1.0.3.tar.gz",
        "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.0.3/bazel-skylib-1.0.3.tar.gz",
    ],
    sha256 = "1c531376ac7e5a180e0237938a2536de0c54d93f5c278634818e0efc952dd56c",
)
load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
bazel_skylib_workspace()


rules_scala_version="116709091e5e1aab3346184217b565f4cb7ba4eb"
http_archive(
     name = "io_bazel_rules_scala",
     strip_prefix = "rules_scala-%s" % rules_scala_version,
     type = "zip",
     url = "https://github.com/bazelbuild/rules_scala/archive/%s.zip" % rules_scala_version,
)
load("@io_bazel_rules_scala//:version.bzl", "bazel_version")
bazel_version(name = "bazel_version")

load("@io_bazel_rules_scala//scala:scala.bzl", "scala_repositories")
scala_repositories()

register_toolchains(
    "//toy:diagnostics_reporter_toolchain"
)

protobuf_tag = "3.13.0.1"
protobuf_sha256 = "f25dbda80787cb5494ace9fe16b3d380a43592e97d5fed31547ef8d331c5253d"
http_archive(
    name = "com_google_protobuf",
    sha256 = protobuf_sha256,
    strip_prefix = "protobuf-{}".format(protobuf_tag),
    type = "zip",
    url = "https://github.com/protocolbuffers/protobuf/archive/v{}.zip".format(protobuf_tag),
)

load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
protobuf_deps()
```

In `toy/BUILD`:

```
load("@io_bazel_rules_scala//scala:scala_toolchain.bzl", "scala_toolchain")
load("@io_bazel_rules_scala//scala:scala.bzl", "scala_library", "scala_binary", "scala_test")

scala_toolchain(
    name = "diagnostics_reporter_toolchain_impl",
    enable_diagnostics_report = True,
    visibility = ["//visibility:public"],
)

toolchain(
    name = "diagnostics_reporter_toolchain",
    toolchain = "diagnostics_reporter_toolchain_impl",
    toolchain_type = "@io_bazel_rules_scala//scala:toolchain_type",
    visibility = ["//visibility:public"],
)

scala_binary(
    name = "main",
    srcs = ["Main.scala"],
    main_class = "me.ex0ns.bazel.Main"
)
```

And in `toy/Main.scala`

```scala
package me.ex0ns.bazel

object Main {

  def main(args: Array[String]): Unit =
    println("Hello world")
}
 ```

I then ran ```coursier launch -r sonatype:snapshots org.jetbrains.bsp:bazel-bsp:0.1.0-SNAPSHOT -M org.jetbrains.bsp.bazel.Install``` at the root of the project.

After that, when importing my project in metals (in VSCode) I could see a lot of output coming from bazel, which is a good sign, a few instant after a project appears in the `Project` tab with a name `//toy:main`

I created the following val:
```scala
val time : Instant = Instant.ofEpochMilli(10000L)
```
And the import of `java.time.Instant` worked !

This looks promising ! I tried to port those changes to a larger (huge) repository build with bazel and it did not work, I'll need to investigate further but I wanted to let you know the result of my first tests with bazel-bsp

**Comment by jtaylor100** [comment]
I have tried to do a configuration similar to @ex0ns:

1. Installing metals using VS Code extensions.
2. Installing bazel-bsp through coursier and adding the toolchain configuration to my ./WORKSPACE and ./BUILD.bazel files (as per the bazel-bsp README).

Things like the `java.time.Instant` import above worked, but the bazel-bsp integration wasn't working since I saw output messages like (paths redacted): 

```
2021.09.09 17:34:47 WARN  Build server is not auto-connectable.
2021.09.09 17:34:47 WARN  the build tool Bazel is not supported by Metals, please open an issue if you would like to contribute to improve the situation.
2021.09.09 17:34:47 WARN  no build target for: /home/jtaylor100/*******/src/main/scala/Adder.scala
2021.09.09 17:34:47 WARN  no build target for: /home/jtaylor100/*******/src/test/scala/ExampleSpec.scala
2021.09.09 17:34:50 INFO  no build target: using presentation compiler with only scala-library: 2.12.14
```

The `java.time.Instant` import was from using the presentation compiler.

So then I took an extra step:

3. Restart the build server within VS Code.

Then I could see bazel was being executed, since I could see lines like:

```
2021.09.09 17:44:15 INFO  Analyzing: 3 targets (0 packages loaded, 0 targets configured)
```

in the VS Code output. However after loading some packages, metals errored with:

````
2021.09.09 17:44:22 ERROR Failed to connect with build server, no functionality will work.
scala.meta.internal.metals.MetalsBspException: BSP connection failed in the attempt to get: ScalacOptionsResult.  java.lang.RuntimeException: java.io.IOException: Cannot run program "/usr/local/bin/bazel": error=7, Argument list too long
	at scala.meta.internal.metals.BuildServerConnection$$anonfun$1.applyOrElse(BuildServerConnection.scala:237)
	at scala.meta.internal.metals.BuildServerConnection$$anonfun$1.applyOrElse(BuildServerConnection.scala:229)
	at scala.concurrent.Future.$anonfun$recoverWith$1(Future.scala:417)
	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:829)
```

Not sure where to go next with debugging the above error. Does anybody have any ideas?

**Comment by ex0ns** [comment]
Do you have a repo with your test ? I can't reproduce it right now, but it might be related to your specific setup (I've never run Windows and  WSL, I've tried the previous method on both Mac OS and Linux)

**Comment by jtaylor100** [comment]
Sorry, the project is proprietary. I'll ping you if I can recreate it in a public repo 👍 Though from the error I'm pretty sure it's to do with the WSL setup.

**Comment by ValdemarGr** [comment]
bazel-bsp just had a 2.0 release and seems to work mostly out of the box for our project with external dependencies, scalapb and macros
https://github.com/JetBrains/bazel-bsp/releases/tag/2.0.0
![image](https://user-images.githubusercontent.com/22582195/165566831-5c5b0639-b04c-47ca-9ff4-19e1572b7c26.png)

**Comment by omerlh** [comment]
I just tested it out, and it worked really well!

**Comment by tgodzik** [comment]
There is probably an issue with references etc. since we rely on semanticdb generation by the build tool. We haven't figured yet out how to do it in Bazel.

**Comment by worace** [comment]
@ValdemarGr could you give any more info on the setup you are using in that example? I think I understand how to launch/install BSP in the bazel project using the command given in that README (`cs launch org.jetbrains.bsp:bazel-bsp:2.2.0 -M org.jetbrains.bsp.bazel.install.Install`)

But I'm unsure then how to "connect" to that setup from your LSP-enabled editor. Is the idea that you simply launch LSP like normal and it detects the existing BSP configuration (maybe also an existing BSP process? I'm unsure about this part) ?

Or is there some other setup required to tell LSP to use the BSP tooling rather than its normal path. Thanks for any advice!

**Comment by ValdemarGr** [comment]
> @ValdemarGr could you give any more info on the setup you are using in that example? I think I understand how to launch/install BSP in the bazel project using the command given in that README (`cs launch org.jetbrains.bsp:bazel-bsp:2.2.0 -M org.jetbrains.bsp.bazel.install.Install`)
> 
> But I'm unsure then how to "connect" to that setup from your LSP-enabled editor. Is the idea that you simply launch LSP like normal and it detects the existing BSP configuration (maybe also an existing BSP process? I'm unsure about this part) ?
> 
> Or is there some other setup required to tell LSP to use the BSP tooling rather than its normal path. Thanks for any advice!

@worace 
I use metals, it picks up the bsp target generated by the command you wrote. After running `cs launch org.jetbrains.bsp:bazel-bsp:2.2.0 -M org.jetbrains.bsp.bazel.install.Install`, try opening the metals logs and then restart the build.

Ensure that your bazel projects adheres to https://github.com/JetBrains/bazel-bsp/blob/master/docs/usage/scala.md

If you also have a bloop configuration on the path it seems that it takes priority; I can change it with `:MetalsSwitchBsp` in vim.

If you see any auth related problems, try disabling all .bazelrc cache related commands (I encountered auth related errors with our remote cache).

**Comment by worace** [comment]
Awesome, thank you. Will try these out. I did get metals connecting to bsp earlier and some features working, but it seemed to be having issues resolving anything with 3rd party deps. Maybe some of these scala configuration tips will help me out.

**Comment by nigredo-tori** [comment]
@ValdemarGr, @omerlh, could you please take a look at this example project: [bazel-example.zip](https://github.com/scalameta/metals-feature-requests/files/9459820/bazel-example.zip), and check whether it works with your setup? For me it compiles fine with `bazel test example`, and installing bazel-bsp doesn't produce errors (see `install-bazel-bsp` script) - but Metals doesn't seem to recognize dependencies. I have this in my metals log:
```
2022.08.31 14:15:37 WARN  Build server is not auto-connectable.
2022.08.31 14:15:37 WARN  the build tool Bazel is not supported by Metals, please open an issue if you would like to contribute to improve the situation.
2022.08.31 14:15:37 WARN  no build target for: /home/nigredo/work/tmp/bazel-example/example/ExampleTest.scala
```

**Comment by tgodzik** [comment]
Looks like you need to have `.bsp/bazel.json` that would indicate to Metals which server to start. Bazel is still not supported standalone, but maybe we could merge a limited support :thinking:

**Comment by nigredo-tori** [comment]
@tgodzik, installing bazel-bsp creates `.bsp/bazelbsp.json`, which should be good enough. However, with your comment I figured something out. The issue is that the build server is not **auto**-connectable - but if we **switch** to it (`lsp-metals-bsp-switch` in Emacs, not sure what the LSP command is), then Metals does connect to it. One wrinkle is that Metals refuses to switch if there's only one BSP server, so you need some other `.bsp/*.json` file with a different `name` field to make it work. I just copied `.bsp/bazelbsp.json` as `.bsp/bazel.json` and changed the `name` inside.

So at the moment I can get Metals to recognize and launch bazel BSP, to recompile changed files, and to show errors. For some reason the errors are not cleaned up, so if I save a file with an error, then I'm stuck with that error in my error list until I restart Metals. And, of course, code navigation does not work.

All in all, I can't say that what we have right now is suitable for actual development. So, as an Emacs user who has to use Bazel for work, I'm SOL, and I **have** to switch to IDEA which seems to be one of the few IDEs that support it. :facepalm:

I'll have some free time in the next couple of weeks. Is there anything I can do to bring Bazel support in Metals closer to reality, in any shape or form? I've thought about writing a Bazel aspect to generate bloop files - maybe reviving https://github.com/stripe-archive/bazel-bloop-exporter, since it's already been done there. This would be a crutch and annoying to use, but it would be better than nothing (that's basically how I've been living with Maven + Metals before this). However, as I understand it, in the long run it would be to add SemanticDB generation in rules_scala (https://github.com/bazelbuild/rules_scala/issues/952) to hopefully unblock the required changes in bazel-bsp - how difficult would that be to implement with zero prior knowledge of either? Could anyone point me towards SemanticDB generation implementation in Bloop or other projects?

**Comment by tgodzik** [comment]
> One wrinkle is that Metals refuses to switch if there's only one BSP server, so you need some other 

Switch command should only be needed if there is more than open entry in .bsp directory, which is why it wouldn't work without it. Metals should however pick up automatically if there is a single entry. Somehow, it was not able to do that :thinking: 

> I've thought about writing a Bazel aspect to generate bloop files - maybe reviving https://github.com/stripe-archive/bazel-bloop-exporter, since it's already been done there.

That's one possibility, however it's a bit complicated and also makes you lose the pros of using bazel. On really large projects Bloop might not work well enough.

There is a WiP PR here https://github.com/scalameta/metals/pull/3233 which might be a good starting point. There is also some pointers towards how to enable semanticdb for your project, which is as you said the biggest obstacle for us.

I thinking that maybe I could improve the heuristics when semanticdb is not available :thinking: so that at least navigation will work properly. References/renames would still need semanticdb to be produced.

**Comment by worace** [comment]
@nigredo-tori thanks for sharing that...it also worked for me, or at least got me a bit further.

Before, this was my process:

* Setup scala toolchain as documented in [bazel-bsp](https://github.com/JetBrains/bazel-bsp/blob/master/docs/usage/scala.md)
* Run BSP Installer via coursier: `cs launch org.jetbrains.bsp:bazel-bsp:2.2.1 -M org.jetbrains.bsp.bazel.install.Install` (generates `.bsp/bazelbsp.json`)
* Start lsp in emacs -- nothing happens...get the previous warning about `Build server is not auto-connectable` etc

So I did your workaround (create a copy of the `bazelbsp.json` with different name), and used `lsp-metals-bsp-switch`

At this point I got some actual build output in my `*lsp-log*` buffer.

Unfortunately the build is not actually working for me now. I think there may be some incompatibilities with the bazel command that is being used by bazel-bsp in this case and how our particular project is configured. I got a long stream of output followed by this exception:

```
scala.meta.internal.metals.MetalsBspException: BSP connection failed in the attempt to get: WorkspaceBuildTargetsResult.  java.nio.file.FileSystemNotFoundException: Provider "bytestream" not installed
```

([Full Trace](https://gist.github.com/worace/238b522e03c3672037046edd3f9b8bd0))

So unfortunately I can't really contribute any more user testing until I sort this out. But wanted to +1 the point about metals not automatically connecting to the bazel-bsp server.

As @tgodzik pointed out

> Switch command should only be needed if there is more than open entry in .bsp directory, which is why it wouldn't work without it. Metals should however pick up automatically if there is a single entry. Somehow, it was not able to do that thinking

Obviously doing this "duplicate the config file and then manually run switch" isn't really a solution, but I wanted to say that I'm at least seeing the same thing you are.
